engines/e_capi.c

```diff
@@ -1460,26 +1461,26 @@ static PCCERT_CONTEXT capi_find_cert(CAPI_CTX * ctx, const char *id,
     }
 }
 
-static CAPI_KEY *capi_get_key(CAPI_CTX * ctx, const TCHAR *contname,
-                              TCHAR *provname, DWORD ptype, DWORD keyspec)
+static CAPI_KEY *capi_get_key(CAPI_CTX *ctx, const WCHAR *contname,
+                              const WCHAR *provname, DWORD ptype,
+                              DWORD keyspec)
 {
     DWORD dwFlags = 0;
     CAPI_KEY *key = OPENSSL_malloc(sizeof(*key));
 
     if (key == NULL)
         return NULL;
     /* If PROV_RSA_AES supported use it instead */
-    if (ptype == PROV_RSA_FULL && use_aes_csp) {
-        provname = NULL;
+    if (ptype == PROV_RSA_FULL && use_aes_csp &&
+        wcscmp(provname, rsa_enh_cspname) == 0) {
+        provname = rsa_aes_cspname;
```

Comment:
> This means that the user can't insist on the enhanced CSP, right? Is that what we want?

Frankly I don't know. I was just fixing specific problem with another commit... But I suppose we can challenge the original commit. So question is if there is downside in forcing "Microsoft Enhanced Cryptographic Provider v1.0" users to "Microsoft Enhanced RSA and AES Cryptographic Provider", right? I suppose that the "undercover" switch is acceptable if former was chosen by default, i.e. without user explicitly asking for it. Formulation means that it's probably appropriate to argue that if user did ask for specific CSP, then the choice should be respected. However! As private key is bound to CSP, user is not actually motivated to choose CSP. I mean user customarily chooses specific key and is expected to accept whatever CSP that comes with it... I'd argue that it goes beyond the scope of this request, i.e. should be handled separately...